Use Variable instead of Tensor in Function.forward

colesbury opened the pull request:
The Tensor and Variable classes are being merged.
autograd.Function.forward is now called on Variables, but with "no-grad"
mode (torch.no_grad()) enabled.

One benefit is that we no longer have to explicitly track shared
storages.

colesbury committed:
Use Variable instead of Tensor in Function.forward

The Tensor and Variable classes are being merged.
autograd.Function.forward is now called on Variables, but with "no-grad"
mode (torch.no_grad()) enabled.

One benefit is that we no longer have to explicitly track shared
storages.

colesbury committed:
Make Fold/Unfold work on Variables

colesbury committed:
Disable test_inplace_flags for now

See issue #4480

yf225 commented:
@pytorchbot retest this please

gchanan commented on test/test_autograd.py:
```diff
@@ -121,11 +121,8 @@ def forward(ctx, tensor1, pyscalar, tensor2):
             @staticmethod
             @once_differentiable
             def backward(ctx, grad_output):
+                self.assertFalse(torch.is_grad_enabled())
```
I thought this would be true in the forward only?  Or is there not really a difference between a forward and once_differentiable backward?

gchanan commented on test/test_autograd.py:
```diff
@@ -1618,20 +1611,22 @@ def test_reentrant(self):
+                    ctx.x = Variable(x.data, requires_grad=True)
+                    ctx.y = Variable(y_data, requires_grad=True)
+                    ctx.output_var = ctx.x * ctx.y
+                return ctx.output_var[:]
```
what happens if you return ctx.output_var?

colesbury replied:
Then the returned value will have the backward be Reenter.backward instead of MulBackward. This test is weird.

apaszke replied:
Don't you have to detach it? I'm not sure what's going on in here

colesbury replied:
Yeah, `return ctx.output_var.detach()` is probably better than `return ctx.output_var[:]` although it ends up having the same effect in this test case.

gchanan commented on test/test_autograd.py:
```diff
@@ -1618,20 +1611,22 @@ def test_reentrant(self):
             @staticmethod
             def backward(ctx, grad_output):
-                ctx.output_var.sum().backward()
+                with torch.enable_grad():
```
similar to above, why do you torch.enable_grad here?

colesbury replied:
`backward()` runs with no-grad if `create_graph` is False because you don't care about higher order derivatives. We could run with `create_graph=True` here (the previous behavior), but that saves more history then necessary. It's a little more clean to just use `enable_grad()` so that the `sum()` calls is differentiable, but not he multiplication on the line below.

apaszke commented on test/test_jit.py:
```diff
@@ -561,6 +561,7 @@ def fn(x):
         torch._C._tracer_exit((y,))
         self.assertExpectedTrace(trace)
 
+    @unittest.expectedFailure
```
What's wrong?

colesbury replied:
Functions are traced twice because `forward()` now operates on Variables. There might be a related in-place bug too. @ezyang is fixing them in #4807

apaszke replied:
Has that been fixed?

apaszke commented on torch/autograd/function.py:
```diff
@@ -200,30 +186,43 @@ def once_differentiable(fn):
-                b"trying to differentiate twice a function that was marked"
-                b"with @once_differentiable")
+        if not requires_grad:
+            return outputs
```
This is a bit worrying for me. I've seen cases where running something with grad disabled, and later reusing those variables in context where grad is enabled triggers some assertions in SavedVariable, which should never fail. I'll open an issue for that

apaszke commented on torch/autograd/function.py:
```diff
@@ -200,30 +186,43 @@ def once_differentiable(fn):
+        def fake_requires_grad(var):
+            if var is not None:
+                var = var.detach()
+                var.requires_grad = True
```
will they become leaves now? (and will accumulate grad)

colesbury replied:
`fake_requires_grad` creates leaf variables, but the thing that is returned is:

 ```
err_fn(*[fake_requires_grad(v) for v in outputs])
```

The application of `err_fn` creates non-leaf Variables. Accumulate grad will never be called because the backward of `err_fn` throws an exception.

apaszke commented on torch/nn/_functions/loss.py:
```diff
@@ -167,7 +167,7 @@ def forward(ctx, input1, input2, y, margin, size_average):
             output = output / y.size(0)
 
         ctx.save_for_backward(input1, input2, y)
-        return input1.new((output,))
```
do we support scalars now?

colesbury replied:
Not really -- scalars are conditionally enabled by a build flag (`WITH_SCALARS=1`) for development.

apaszke replied:
So how does this work now? Isn't output a Python scalar in here?

colesbury replied:
It's a 1-element 1-d tensor.

apaszke replied:
ohhh because those are Variables so they never degenerate into scalars. makes sense

apaszke reviewed (commented):
Haven't reviewed `python_function.cpp` yet. Will do that later today

colesbury committed:
Use detach() instead of indexing

apaszke commented on torch/autograd/function.py:
```diff
@@ -38,24 +39,9 @@ def mark_dirty(self, *args):
-        indexing and transpose ops.
-        """
-        self.shared_pairs = pairs
+        warnings.warn(
```
I just realized that we won't be able to track sharing if someone uses `set_` (which is not an error on Variables that don't require grad, right), correct? We might want to keep it after all

colesbury replied:
That's true. I don't think `mark_shared_storage` is a good solution though. I think we should be working towards an API that matches the underlying semantics instead of this sort of inversion of control.

In this case, that would mean exposing a `Variable` constructor that takes a `base`/`view` or fixing `set_`.

apaszke replied:
Alternatively we could attach version counters to storages and get sharing tracking completely for free

colesbury replied:
Yeah. Along those lines we could just replace Storage with a 1-d Tensor (at least in Python).

apaszke replied:
idk, storages are kind of nice. Can you add a note that says that we don't track `set_`?

apaszke commented on torch/backends/cudnn/rnn.py:
```diff
@@ -184,7 +184,7 @@ def _copyParams(params_from, params_to):
         for param_from, param_to in zip(layer_params_from, layer_params_to):
             assert param_from.type() == param_to.type()
-            param_to.copy_(param_from, broadcast=False)
+            param_to[:] = param_from.view_as(param_to)
```
any reason why `copy_` became incorrect in here?

colesbury replied:
We don't support `broadcast=False` on Variables.

colesbury replied:
This could just be:

```
param_to.copy_(param_from.view_as(param_to)
```

apaszke commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -310,38 +301,31 @@ using t2var_type = std::unordered_map<PyObject *, THPVariable *>;
 {
   // Increase versions of modified tensors
-  if (!self->dirty_tensors) return;
+  std::unordered_set<PyObject*> dirty_inputs;
```
NB it's probably faster to use a vector for this purpose. I don't think anyone uses function that mutate a lot of inputs

apaszke commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -366,100 +349,82 @@ static void _wrap_outputs(THPFunction *self, t2var_type &t2var,
   };
 
   // Wraps an output Tensor in a Variable or returns the previous wrapper in
   // the case of in-place modification.
```
Comment is out of date

apaszke commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -366,100 +349,82 @@ static void _wrap_outputs(THPFunction *self, t2var_type &t2var,
+    } else if (is_input) {
+      // An input has been returned, but it wasn't modified. Return it as a view
+      // so that we can attach a new grad_fn to the Variable.
+      // return make_variable_view(std::move(prev), std::move(data), output_nr, cdata);
```
should this be a part of this comment?

colesbury replied:
Deleted -- dead code from a previous attempt

apaszke commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -366,100 +349,82 @@ static void _wrap_outputs(THPFunction *self, t2var_type &t2var,
       }
       if (cdata) {
-        prev.rebase_history(output_nr, cdata);
+        var.rebase_history(output_nr, cdata);
```
`rebase_history` has a special implementation for views, right?

colesbury replied:
Yes. For views it will modify the `grad_fn` on the base (using `cdata` here). The `grad_fn` on the view reverses the viewing operation.

On non-views, it just sets `var.output_nr=output_nr` and `var.grad_fn=cdata`

apaszke commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -366,100 +349,82 @@ static void _wrap_outputs(THPFunction *self, t2var_type &t2var,
   };
 
-  t2var_type output2var;
+  std::unordered_set<PyObject*> seen;
```
Is this even used anywhere? Looks like it should be

colesbury replied:
No -- dead code. We no longer have the issue of identity no being preserved because we don't unpack the Variables into Tensors.

apaszke commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -471,71 +436,26 @@ static void _save_variables(THPFunction* self, t2var_type &t2var)
+      bool is_output = variable->cdata.grad_fn().get() == cdata_ptr;
+      self->saved_variables.emplace_back(variable->cdata, is_output);
+    } else if (THPModule_isTensor(obj)) {
+      // TODO: remove once Variable and Tensor classes are merged
```
Can we make something like `// TODO [VarTensor]` so that we have a single thing to grep for after we merge them? There are a few places in this file that could be simplified, and they have a similar comment, but it will be hard to find them later

colesbury replied:
We already have a lot of comments with this pattern. I don't think adding this convention will make it any easier to find those.

colesbury committed:
Remove dead code and comments

apaszke commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -366,100 +349,79 @@ static void _wrap_outputs(THPFunction *self, t2var_type &t2var,
+    bool is_non_differentiable = non_differentiable.count(obj) > 0;
+
+    auto var = as_variable(obj, i);
+    set_history(var, i, is_input, is_modified, is_non_differentiable);
```
Can you tell me why is it the case that we no longer need to track duplication? We will just destroy the previous history assignment in here. In general dealing with repeated outputs is hard, but I think the first one should be a proper Variable, and all other ones should be implicit views, so they can get gradient too. Right now one of `grad_outputs` for a repeated output is guaranteed to be None

colesbury replied:
I think we should prefer returning the same object when possible. I've added explicit tests for this, although other tests triggered this code path too.

When var is a view, this raises an exception because we don't support in-place modifications to views from functions that return multiple values.

Otherwise, the last call to set_history wins. All that will change is the `output_nr` which doesn't really matter.

apaszke replied:
Hmm idk, forbiding that would be safer. Do we depend on it anywhere outside of our test cases? Also, can you add a comment saying that it's a deliberate decision?

colesbury committed:
Add tests for duplicate returns

colesbury commented:
@apaszke can you take another look?

apaszke commented:
Sure, I'll do it later today

apaszke commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -310,38 +301,31 @@ using t2var_type = std::unordered_map<PyObject *, THPVariable *>;
 // into dirty_inputs.  Also does some sanity checking.
-static void _mark_dirty(THPFunction *self, t2var_type &t2var,
-        std::unordered_set<PyObject *> &dirty_inputs)
+static std::vector<PyObject*> _parse_dirty_inputs(THPFunction *self)
```
nit: I think `_mark_dirty` would still be a better name, because it implies that it has a side effect, while `parse_dirty_inputs` would imply that it doesn't.

apaszke commented on torch/cuda/comm.py:
```diff
@@ -191,10 +190,13 @@ def gather(tensors, dim=0, destination=None):
         destination = torch.cuda.current_device()
     if destination == -1:
-        result = getattr(torch, type(tensors[0]).__name__)(expected_size)
+        if isinstance(tensors[0], torch.autograd.Variable):
```
TODO: remove after merging?

apaszke reviewed (approved):
LGTM. Very minor suggestions only

colesbury committed:
Split tracing of Python functions into two parts

Fixes handling of in-place autograd functions where operations are
traced twice due to forward operating on Variables.

Re-enable test_inplace_flags

colesbury commented:
@ezyang can you review the "Split tracing of Python functions into two parts" commit?

colesbury committed:
Wrap tensor outputs before calling mark_non_differentiable

colesbury committed:
Add comments and rename _parse_dirty_inputs

colesbury commented:
@pytorchbot retest this please

ezyang commented on torch/csrc/jit/tracer.cpp:
```diff
@@ -154,6 +154,42 @@ PreTraceInfo preRecordTrace(std::string op, // TODO: make this a Symbol
   return info;  // RVO
 }
 
+PreTraceInfo preRecordPythonTrace(THPObjectPtr pyobj,
```
I guess we should mention that this is copy pasted from above so people remember to update both.

colesbury replied:
I refactored out the common code in preRecordTrace and preRecordPythonTrace

colesbury committed:
Refactor out common code in preRecordTrace and preRecordPythonTrace

ezyang commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -571,53 +565,37 @@ static void _trace_create(PyObject* op_obj, THPFunction* bw_obj,
-    value_traces.emplace_back(tracer::getValueTrace(tracing_state, i));
-    var_flags.push_back(VariableFlags::of(i));
-  }
+  Py_INCREF(op_obj);
```
Do you want me to double check refcounting changes?

colesbury replied:
I'm pretty confident about this one. It corresponds to the `THPObjectPtr()` construction below.

We could add a "retain" argument to the THPObjectPtr constructor like we do in `at::Tensor` to avoid the separate incref call.

ezyang commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -627,13 +605,15 @@ static void _trace_create(PyObject* op_obj, THPFunction* bw_obj,
   // tracing_state->in_eval_subgraph (it's always false, because they are never part of backward
   // subgraphs AND we don't even materialize the forward function).
   if (!passes_state_transparently) {
+    // TODO: sgross and ezyang don't know if this is right
```
Paging @apaszke

apaszke replied:
Idk, looks fine. What's wong?
